src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/CompiledRegexRunner.cs

```diff
@@ -1,23 +1,172 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+
 namespace System.Text.RegularExpressions
 {
     internal sealed class CompiledRegexRunner : RegexRunner
     {
-        private readonly Action<RegexRunner> _goMethod;
-        private readonly Func<RegexRunner, bool> _findFirstCharMethod;
+        internal delegate void goDelegate(RegexRunner runner, ReadOnlySpan<char> inputSpan);
+        internal delegate bool findFirstCharDelegate(RegexRunner runner, ReadOnlySpan<char> inputSpan);
+
+        private readonly goDelegate _goMethod;
+        private readonly findFirstCharDelegate _findFirstCharMethod;
 
-        public CompiledRegexRunner(Action<RegexRunner> go, Func<RegexRunner, bool> findFirstChar, int trackCount)
+        public CompiledRegexRunner(goDelegate go, findFirstCharDelegate findFirstChar, int trackCount)
```

Comment:
We should have a single delegate here to a ScanDelegate and which the Scan override calls.  The whole body of what you have in the Scan override below should instead be in the code the ILCompiler generates for the ScanDelegate, just as it will need to be for the source generator.  You can still generate two separate methods for Go and FindFirstChar (at least for now), and the Scan override you generate will call them.  So basically we end up with a single virtual call to Scan, which in turn makes a single ScanDelegate delegate invocation, which calls the generated IL implementation that loops around non-virtual calls to FindFirstChar and Go.  Eventually we can then combine FindFirstChar and Go into the single generated Scan method rather than generating three separate methods.